backend/corpora/dataset_processing/upload.py

```diff
@@ -0,0 +1,139 @@
+import logging
+import threading
+import queue
+import requests
+
+from ..common.corpora_orm import DbDatasetProcessingStatus, UploadStatus
+from ..common.entities import Dataset
+from ..common.utils.db_utils import db_session_manager
+from ..common.utils.math_utils import MB
+
+logger = logging.getLogger(__name__)
+
+
+class ProgressTracker:
+    def __init__(self, file_size: int):
+        self.file_size: int = file_size
+        self._progress: int = 0
+        self.progress_lock: threading.Lock = threading.Lock()  # prevent concurrent access of _progress
+        self.stop_updater: threading.Event = threading.Event()  # Stops the update_progress thread
+        self.stop_uploader: threading.Event = threading.Event()  # Stops the uploader threads
+        self.error: queue.Queue = queue.Queue(maxsize=1)  # Track errors
```

Comment:
Why is this a queue?

Reply:
I thought it was easier than creating another lock to prevent concurrent access.

Reply:
also if multiple errors occur we want to catch them.